File: 24-Graph-Part-II/cpp/01-kosarajus-algorithm-maximum-number-of-non-overlapping-substrings.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<string> maxNumOfSubstrings(string s) {
        int n = s.length();
        vector<int> left(26, n), right(26, -1);
        for (int i = 0; i < n; ++i) {
            left[s[i] - 'a'] = min(left[s[i] - 'a'], i);
            right[s[i] - 'a'] = i;
        }

        vector<string> res;
        int r = -1;

        for (int i = 0; i < n; ++i) {
            if (i != left[s[i] - 'a']) continue;
            int newR = right[s[i] - 'a'];
            for (int j = i + 1; j <= newR; ++j) {
                if (left[s[j] - 'a'] < i) {
                    newR = n;
                    break;
                }
                newR = max(newR, right[s[j] - 'a']);
            }
            if (newR < n && (i > r || newR < right[s[r] - 'a'])) {
                if (i > r) res.push_back(s.substr(i, newR - i + 1));
                else res.back() = s.substr(i, newR - i + 1);
                r = newR;
            }
        }
        return res;
    }
};

File: 24-Graph-Part-II/python/01_kosarajus_algorithm_maximum_number_of_non_overlapping_substrings.py
from typing import List

class Solution:
    def maxNumOfSubstrings(self, s: str) -> List[str]:
        n = len(s)
        left = [n] * 26
        right = [-1] * 26

        for i, ch in enumerate(s):
            idx = ord(ch) - ord('a')
            left[idx] = min(left[idx], i)
            right[idx] = i

        res = []
        r = -1

        for i in range(n):
            idx = ord(s[i]) - ord('a')
            if i != left[idx]:
                continue
            newR = right[idx]
            valid = True
            j = i + 1
            while j <= newR:
                j_idx = ord(s[j]) - ord('a')
                if left[j_idx] < i:
                    valid = False
                    break
                newR = max(newR, right[j_idx])
                j += 1

            if valid and newR < n and (i > r or newR < right[ord(s[r]) - ord('a')]):
                if i > r:
                    res.append(s[i:newR + 1])
                else:
                    res[-1] = s[i:newR + 1]
                r = newR

        return res
